### PR TITLE
Zero2: avoid graph breaks in torch.compile by using param_idx

### PR DESCRIPTION
**Pull request by nelyahu**
inside reduce_independent_p_g_buckets_and_remove_grads and in reduce_ipg_grads which are being executed during the BWD hook in zero2, the model param is being stored inside params_in_ipg_bucket. torch.compile has hard time tracing parameters.
By using the param's static index inside the group the same logic can be maintain with less complexity.

### CODE REVIEW

File: deepspeed/runtime/zero/stage_1_and_2.py [hunk 1]
```diff
@@ -1067,7 +1068,7 @@ def average_tensor(self, tensor):
 
             process_group = self.dp_process_group
             # count = 0
-            for i, param, param_id in self.params_in_ipg_bucket:
+            for i, param_idx_in_group, param_id in self.params_in_ipg_bucket:
```
Comment (tjruwase) [inline deepspeed/runtime/zero/stage_1_and_2.py:1071]:
```suggestion
            for i, param_idx_in_group, param_id in self.params_in_ipg_bucket:
                 param = self.bit16_groups[group_idx][param_idx_in_group]
```

Comment (tjruwase) [reply deepspeed/runtime/zero/stage_1_and_2.py:1071]: @nelyahu, do you have any thoughts on this suggestion?

Comment (nelyahu) [reply deepspeed/runtime/zero/stage_1_and_2.py:1071]: @tjruwase Done. sorry for the delay

**Review by tjruwase — APPROVED** [review]

File: deepspeed/runtime/zero/stage_1_and_2.py [hunk 2]
```diff
@@ -1067,7 +1068,8 @@ def average_tensor(self, tensor):
...
+                param = self.bit16_groups[group_idx][param_idx_in_group]
```
Comment (loadams) [inline deepspeed/runtime/zero/stage_1_and_2.py:1072]: @nelyahu - this is failing on group_idx just fyi

Comment (nelyahu) [reply deepspeed/runtime/zero/stage_1_and_2.py]: fixed

**Pull request merged**